File: scripts/fetch_dataset.py
import os
import sys
import json
import time
import urllib.request
import urllib.parse
import urllib.error
from dotenv import load_dotenv

load_dotenv(dotenv_path='.env.local')

BASE_URL = os.getenv('API_BASE_URL', 'https://solve.ivy.homes').rstrip('/')
API_KEY = os.getenv('API_KEY', '')
LOGIN_EMAIL = os.getenv('LOGIN_EMAIL', 'demo1@ivy.homes')
LOGIN_PASSWORD = os.getenv('LOGIN_PASSWORD', '')

os.makedirs('data', exist_ok=True)

def get_token():
    url = f"{BASE_URL}/auth/login"
    body = json.dumps({"email": LOGIN_EMAIL, "password": LOGIN_PASSWORD}).encode('utf-8')
    headers = {'Content-Type': 'application/json', 'X-API-Key': API_KEY}
    req = urllib.request.Request(url, data=body, headers=headers, method='POST')
    with urllib.request.urlopen(req) as resp:
        data = json.loads(resp.read().decode('utf-8'))
        return data['access_token']

token = get_token()
print("Token acquired successfully!")

def fetch_all(endpoint_name, path):
    global token
    print(f"\n--- Fetching complete dataset for {endpoint_name} ({path}) ---")
    all_records = []
    offset = 0
    limit = 50
    total = None
    
    while True:
        url = f"{BASE_URL}{path}?offset={offset}&limit={limit}"
        headers = {'X-API-Key': API_KEY, 'Authorization': f'Bearer {token}'}
        req = urllib.request.Request(url, headers=headers)
        
        try:
            with urllib.request.urlopen(req) as resp:
                data = json.loads(resp.read().decode('utf-8'))
                results = data.get('results', [])
                total = data.get('total')
                has_more = data.get('has_more', False)
                all_records.extend(results)
                print(f"Offset {offset:4d}: fetched {len(results)} items (Total accumulated: {len(all_records)} / {total})", flush=True)
                
                if not has_more or len(results) == 0:
                    break
                offset += len(results)
        except urllib.error.HTTPError as e:
            if e.code == 401:
                print("Token expired! Re-authenticating...")
                token = get_token()
                continue
            else:
                print(f"HTTPError {e.code}: {e.read().decode()}")
                break
        
        # Fast fetching within rate limit
        
    out_file = f"data/{endpoint_name}.json"
    with open(out_file, 'w', encoding='utf-8') as f:
        json.dump(all_records, f, indent=2)
    print(f"Saved {len(all_records)} records to {out_file}")
    return all_records

listings = fetch_all('listings', '/v1/listings')
rentals = fetch_all('rentals', '/v1/rentals')
projects = fetch_all('projects', '/v1/projects')

print("\n=== DATASET DOWNLOAD COMPLETE ===")
print(f"Total Listings: {len(listings)}")
print(f"Total Rentals:  {len(rentals)}")
print(f"Total Projects: {len(projects)}")

File: scripts/fetch_dataset.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const rootDir = path.resolve(__dirname, '..');

function loadEnv() {
  const envPath = path.join(rootDir, '.env.local');
  if (fs.existsSync(envPath)) {
    const content = fs.readFileSync(envPath, 'utf-8');
    for (const line of content.split('\n')) {
      const trimmed = line.trim();
      if (trimmed && !trimmed.startsWith('#') && trimmed.includes('=')) {
        const idx = trimmed.indexOf('=');
        const key = trimmed.slice(0, idx).trim();
        let val = trimmed.slice(idx + 1).trim();
        if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
          val = val.slice(1, -1);
        }
        if (!process.env[key]) {
          process.env[key] = val;
        }
      }
    }
  }
}
loadEnv();

const BASE_URL = (process.env.VITE_API_BASE_URL || process.env.API_BASE_URL || 'https://solve.ivy.homes').replace(/\/+$/, '');
const API_KEY = process.env.VITE_API_KEY || process.env.API_KEY || 'IVY26-05FB73DA5767';
const LOGIN_EMAIL = process.env.VITE_LOGIN_EMAIL || process.env.LOGIN_EMAIL || 'demo1@ivy.homes';
const LOGIN_PASSWORD = process.env.VITE_LOGIN_PASSWORD || process.env.LOGIN_PASSWORD || 'password123';

const dataDir = path.join(rootDir, 'data');
if (!fs.existsSync(dataDir)) {
  fs.mkdirSync(dataDir, { recursive: true });
}

async function getToken() {
  const url = `${BASE_URL}/auth/login`;
  const res = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-API-Key': API_KEY
    },
    body: JSON.stringify({ email: LOGIN_EMAIL, password: LOGIN_PASSWORD })
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Login failed with status ${res.status}: ${text}`);
  }
  const data = await res.json();
  return data.access_token;
}

let token = await getToken();
console.log('Token acquired successfully!');

async function fetchAll(endpointName, apiPath) {
  console.log(`\n--- Fetching complete dataset for ${endpointName} (${apiPath}) ---`);
  const allRecords = [];
  let offset = 0;
  const limit = 50;

  while (true) {
    const url = `${BASE_URL}${apiPath}?offset=${offset}&limit=${limit}`;
    const res = await fetch(url, {
      headers: {
        'X-API-Key': API_KEY,
        'Authorization': `Bearer ${token}`
      }
    });

    if (res.status === 401) {
      console.log('Token expired! Re-authenticating...');
      token = await getToken();
      continue;
    }

    if (!res.ok) {
      const text = await res.text();
      console.error(`HTTP Error ${res.status}: ${text}`);
      break;
    }

    const data = await res.json();
    const results = data.results || [];
    const total = data.total;
    const hasMore = data.has_more || false;

    allRecords.push(...results);
    console.log(`Offset ${String(offset).padStart(4, ' ')}: fetched ${results.length} items (Total accumulated: ${allRecords.length} / ${total})`);

    if (!hasMore || results.length === 0) {
      break;
    }
    offset += results.length;
  }

  const outFile = path.join(dataDir, `${endpointName}.json`);
  fs.writeFileSync(outFile, JSON.stringify(allRecords, null, 2), 'utf-8');
  console.log(`Saved ${allRecords.length} records to ${outFile}`);
  return allRecords;
}

const listings = await fetchAll('listings', '/v1/listings');
const rentals = await fetchAll('rentals', '/v1/rentals');
const projects = await fetchAll('projects', '/v1/projects');

console.log('\n=== DATASET DOWNLOAD COMPLETE ===');
console.log(`Total Listings: ${listings.length}`);
console.log(`Total Rentals:  ${rentals.length}`);
console.log(`Total Projects: ${projects.length}`);
